pkg/agent/memberlist/cluster_test.go

```diff
@@ -0,0 +1,772 @@
+			name:                     "update externalIPPool with the matched nodeSelector then local node should selected",
+			expectEgressSelectResult: true,
+			newEIPnodeSelectors:      metav1.LabelSelector{MatchLabels: map[string]string{"env": "test"}},
+		},
+		{
+			name:                     "update externalIPPool with nil nodeSelector then local node should selected",
+			expectEgressSelectResult: true,
+			newEIPnodeSelectors:      metav1.LabelSelector{},
+		},
+		{
+			name:                     "update externalIPPool refresh back then local node should not selected",
+			expectEgressSelectResult: false,
+			newEIPnodeSelectors:      eip.Spec.NodeSelector,
+		},
+	}
+	for _, tCase := range testCaseEIPEvecnts {
+		t.Run(tCase.name, func(t *testing.T) {
+			eip.Spec.NodeSelector = tCase.newEIPnodeSelectors
+			_, err := crdClient.CrdV1alpha2().ExternalIPPools().Update(context.TODO(), eip, metav1.UpdateOptions{})
+			if err != nil {
+				t.Fatalf("Update externalIPPool error: %v", err)
+			}
+			assert.NoError(t, wait.Poll(time.Millisecond*100, time.Second, func() (done bool, err error) {
+				newEIP, _ := crdClient.CrdV1alpha2().ExternalIPPools().Get(context.TODO(), eip.Name, metav1.GetOptions{})
+				return reflect.DeepEqual(eip, newEIP), nil
+			}))
+			res := s.ShouldSelectEgress(eg)
+			assert.Equal(t, tCase.expectEgressSelectResult, res, "select node for egress result not match")
+		})
+	}
+
+	eipEnvTest := &crdv1a2.ExternalIPPool{
+		TypeMeta: metav1.TypeMeta{
+			Kind: "CustomResourceDefinition",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "eipEnvTest",
+		},
+		Spec: crdv1a2.ExternalIPPoolSpec{NodeSelector: metav1.LabelSelector{MatchLabels: map[string]string{"env": "test"}}},
+	}
+	// test create new eip
+	egEnvTest := &crdv1a2.Egress{
+		ObjectMeta: metav1.ObjectMeta{Name: "testClusterEgress1", UID: "fakeUID1"},
+		Spec:       crdv1a2.EgressSpec{ExternalIPPool: eipEnvTest.Name},
+	}
+	createAndCheckEIP(eipEnvTest)
+	assert.Equal(t, true, s.ShouldSelectEgress(egEnvTest), "select node for egress result not match")
+	assert.Equal(t, false, s.ShouldSelectEgress(eg), "select node for egress result not match")
+
+	// test delete eip
+	deleteAndCheckEIP := func(eipName string) {
+		err := crdClient.CrdV1alpha2().ExternalIPPools().Delete(context.TODO(), eipName, metav1.DeleteOptions{})
+		if err != nil {
+			t.Fatalf("Delete externalIPPool error: %v", err)
+		}
+		assert.NoError(t, wait.Poll(time.Millisecond*100, time.Second, func() (done bool, err error) {
+			newEIP, _ := crdClient.CrdV1alpha2().ExternalIPPools().Get(context.TODO(), eipName, metav1.GetOptions{})
+			return nil == newEIP, nil
+		}))
+	}
+	// eip nodeSelector is nil
+	deleteAndCheckEIP(eipEnvTest.Name)
+	assert.Equal(t, false, s.ShouldSelectEgress(eg), "select node for egress result not match")
+	assert.Equal(t, false, s.ShouldSelectEgress(egEnvTest), "select node for egress result not match")
+
+}
+
+func TestCluster_ShouldSelectNodeFailedOrAddedBySortHash(t *testing.T) {
```

Comment:
This can explain why consistentHashMap is prefered but it does not make sense to include in the final code as it's never used and can confuse people. It's not helpful for the coverage too. Maybe you can just add the data to the comment of consistentHashMap's usage.

Reply:
done